Fixest vcov catches

grantmcdermott opened the pull request:
I know you (we!) largely came out against hardcoding specific model classes in #280. But been I've running into this whilst teaching, so thought I might as well have a go at it. I won't be offended if you decide not to merge given the aforementioned stance, but I _do_ think the current behaviour is a little sub-optimal given the popularity of fixest. 

(Note: If you're game, I can probably do something similar for lfe::felm, which also requires a couple of control flow catches. estimatr is harder b/c --- unlike lfe::felm --- it doesn't include the iid vcov errors as part of the return object).

cc @elbersb 

Examples:.

``` r
devtools::load_all('~/Documents/Projects/modelsummary/')
#> ℹ Loading modelsummary
library(fixest)

mod_lm =         lm(hp ~ mpg + drat, mtcars)
mod_feols =   feols(hp ~ mpg + drat, mtcars, vcov = ~vs)

mods = list('lm' = mod_lm, 'feols' = mod_feols)

msummary(mods, gof_omit = 'R2|IC|Log|F', output = 'markdown')
```

|             |    lm    |     feols      |
|:------------|:--------:|:--------------:|
| (Intercept) | 278.515  |    278.515     |
|             | (55.415) |    (20.573)    |
| mpg         |  -9.985  |     -9.985     |
|             | (1.792)  |    (4.110)     |
| drat        |  19.126  |     19.126     |
|             | (20.198) |    (32.644)    |
| Num.Obs.    |    32    |       32       |
| Std.Errors  |          | Clustered (vs) |

``` r
msummary(mods, vcov = 'iid', gof_omit = 'R2|IC|Log|F', output = 'markdown') 
```

|             |    lm    |  feols   |
|:------------|:--------:|:--------:|
| (Intercept) | 278.515  | 278.515  |
|             | (55.415) | (55.415) |
| mpg         |  -9.985  |  -9.985  |
|             | (1.792)  | (1.792)  |
| drat        |  19.126  |  19.126  |
|             | (20.198) | (20.198) |
| Num.Obs.    |    32    |    32    |
| Std.Errors  |   IID    |   IID    |

``` r
## Aside: this "doesn't" work...
msummary(mods, vcov = c('iid', NULL), gof_omit = 'R2|IC|Log|F', output = 'markdown')
```

|             |    lm    |  feols   |
|:------------|:--------:|:--------:|
| (Intercept) | 278.515  | 278.515  |
|             | (55.415) | (55.415) |
| mpg         |  -9.985  |  -9.985  |
|             | (1.792)  | (1.792)  |
| drat        |  19.126  |  19.126  |
|             | (20.198) | (20.198) |
| Num.Obs.    |    32    |    32    |
| Std.Errors  |   IID    |   IID    |

``` r
## ... but only b/c the vcov list drops the NULL element (as it should)
c('iid', NULL)
#> [1] "iid"

msummary(mods, vcov = 'robust', gof_omit = 'R2|IC|Log|F', output = 'markdown') 
```

|             |    lm    |  feols   |
|:------------|:--------:|:--------:|
| (Intercept) | 278.515  | 278.515  |
|             | (46.992) | (46.992) |
| mpg         |  -9.985  |  -9.985  |
|             | (2.471)  | (2.471)  |
| drat        |  19.126  |  19.126  |
|             | (23.133) | (23.133) |
| Num.Obs.    |    32    |    32    |
| Std.Errors  |  Robust  |  Robust  |

``` r
msummary(mods, vcov = 'HC1', gof_omit = 'R2|IC|Log|F', output = 'markdown')
```

|             |    lm    |  feols   |
|:------------|:--------:|:--------:|
| (Intercept) | 278.515  | 278.515  |
|             | (42.037) | (42.037) |
| mpg         |  -9.985  |  -9.985  |
|             | (2.209)  | (2.209)  |
| drat        |  19.126  |  19.126  |
|             | (20.572) | (20.572) |
| Num.Obs.    |    32    |    32    |
| Std.Errors  |   HC1    |   HC1    |

``` r
msummary(mods, vcov = ~ vs, gof_omit = 'R2|IC|Log|F', output = 'markdown')
```

|             |    lm    |  feols   |
|:------------|:--------:|:--------:|
| (Intercept) | 278.515  | 278.515  |
|             | (20.573) | (20.573) |
| mpg         |  -9.985  |  -9.985  |
|             | (4.110)  | (4.110)  |
| drat        |  19.126  |  19.126  |
|             | (32.644) | (32.644) |
| Num.Obs.    |    32    |    32    |
| Std.Errors  |  C: vs   |  C: vs   |

``` r
msummary(mods, vcov = ~ cyl, gof_omit = 'R2|IC|Log|F', output = 'markdown')
```

|             |    lm    |  feols   |
|:------------|:--------:|:--------:|
| (Intercept) | 278.515  | 278.515  |
|             | (40.571) | (40.571) |
| mpg         |  -9.985  |  -9.985  |
|             | (3.571)  | (3.571)  |
| drat        |  19.126  |  19.126  |
|             | (34.092) | (34.092) |
| Num.Obs.    |    32    |    32    |
| Std.Errors  |  C: cyl  |  C: cyl  |

<sup>Created on 2021-08-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

vincentarelbundock commented on R/get_vcov.R:
```diff
@@ -15,6 +15,42 @@ get_vcov <- function(model, vcov = NULL, conf_level = NULL, ...) {
     "function that returns one of these, such as `stats::vcov`.")
 
+  ## fixest models
+  if (class(model) %in% c("fixest", "fixest_multi")) {
```
My guess is the test failures are due to this line, because `class(model)` is a vector with two elements for these classes. So I think you need `class(model)[1] %in% ...` or maybe `inherits(model, "fixest") || inherits(model, "fixest_multi")`

vincentarelbundock commented:
This looks fantastic, thanks! I'd be happy to merge when tests pass. I'm pretty sure I identified the culprit in my comment above, and it should be a very easy fix.

grantmcdermott commented:
I updated with your hotfix and also added some better checking + flow, depending on the fixest version. Looks like everything is passing except for an annoying warning about an undeclared dependency (`Matrix`?) for `lfe`, which I added to the "hardcoded"  test suite. Happy to remove for now if that's easiest.

vincentarelbundock commented:
Fantastic!
